WordPress/src/main/java/org/wordpress/android/util/encryptedlogging/EncryptedLogWriter.kt

```diff
@@ -0,0 +1,95 @@
+package org.wordpress.android.util.encryptedlogging
+
+import android.util.Base64
+import com.goterl.lazycode.lazysodium.interfaces.SecretStream
+import com.goterl.lazycode.lazysodium.utils.Key
+import java.io.File
+import java.io.FileWriter
+import java.util.UUID
+
+/**
+ * EncryptedLogWriter creates encrypted logs.
+ **
+ * @param file A file object representing the log file destination.
+ * @param publicKey The public key used to encrypt the log.
+ * @param uuid (Optional) The UUID string associated with the encrypted log. If one is not provided, it is generated for you.
+ * @constructor Creates the `EncryptedLogWriter`. Call `write` to append data, and `close` when to write out the postamble.
+ *
+ */
+class EncryptedLogWriter(
+    val file: File,
+    publicKey: Key,
+    private val uuid: String = UUID.randomUUID().toString()
+) {
+    private val fileWriter: FileWriter = FileWriter(file)
+    private val sodium = EncryptionUtils.sodium
+    private var state = SecretStream.State.ByReference()
+
+    init {
+        var header = ByteArray(SecretStream.HEADERBYTES)
+        val key = SecretStreamKey.generate()
+        check(sodium.cryptoSecretStreamInitPush(state, header, key.bytes))
+        writeHeader(key.encrypt(publicKey), header)
+    }
+
+    /**
+     * Encrypt and write the provided string to the encrypted log file.
+     * @param String: The string to be written to the file.
+     */
+    fun write(string: String) {
+        val encryptedString = encryptMessage(string, SecretStream.TAG_MESSAGE)
+        fileWriter.write("\t\t\"$encryptedString\",\n")
+    }
+
+    /**
+     * Add the closing file tag, and ensure that all buffers are flushed to disk.
+     */
+    fun close() {
+        val encryptedClosingTag = encryptMessage("", SecretStream.TAG_FINAL)
+        fileWriter.write("\t\t\"$encryptedClosingTag\"\n")
+        fileWriter.write("\t]\n")
+        fileWriter.write("}")
+        fileWriter.flush()
+    }
+
+    // An internal convenience function to push more data into the sodium secret stream.
+    private fun encryptMessage(string: String, tag: Byte): String {
+        val plainBytes = string.toByteArray()
+
+        val encryptedBytes = ByteArray(SecretStream.ABYTES + plainBytes.size) // Stores the encrypted bytes
+        check(sodium.cryptoSecretStreamPush(this.state, encryptedBytes, plainBytes, plainBytes.size.toLong(), tag)) {
+            "Unable to encrypt message: $string"
+        }
+
+        return base64Encode(encryptedBytes)
+    }
+
+    // An internal convenience function to extract the header writing process.
+    private fun writeHeader(key: EncryptedSecretStreamKey, header: ByteArray) {
+        check(SecretStream.Checker.headerCheck(header.size))
+
+        val encodedEncryptedKey = base64Encode(key.bytes)
+        check(encodedEncryptedKey.length == 108) {
```

Comment:
This is a somewhat "magic" number – it's the right number for this implementation, and is enforced by a unit test, but having the check here helps debug implementation issues (for instance, if a developer tries to use the wrong kind of `KeyPair` with this class).

It's a constant constant – ie, this will never change as long as this implementation exists, so it shouldn't be extracted to a class constant – if a developer changes this value, they break the compatibility with the server-side decryption logic, which is why it's directly in the code like this.

Reply:
If there is ever a time to use numbers directly, I think this would be it. I really like the `checks` all over and the message for that already explains the "magic" bit. (kind of) 👍